src/Storages/System/StorageSystemDetachedParts.cpp

```diff
@@ -16,18 +17,54 @@ StorageSystemDetachedParts::StorageSystemDetachedParts(const StorageID & table_i
 {
     StorageInMemoryMetadata storage_metadata;
     storage_metadata.setColumns(ColumnsDescription{{
-        {"database", std::make_shared<DataTypeString>()},
-        {"table", std::make_shared<DataTypeString>()},
-        {"partition_id", std::make_shared<DataTypeNullable>(std::make_shared<DataTypeString>())},
-        {"name", std::make_shared<DataTypeString>()},
-        {"disk", std::make_shared<DataTypeString>()},
-        {"reason", std::make_shared<DataTypeNullable>(std::make_shared<DataTypeString>())},
+        {"database",         std::make_shared<DataTypeString>()},
+        {"table",            std::make_shared<DataTypeString>()},
+        {"partition_id",     std::make_shared<DataTypeNullable>(std::make_shared<DataTypeString>())},
+        {"name",             std::make_shared<DataTypeString>()},
+        {"bytes_on_disk",    std::make_shared<DataTypeUInt64>()},
+        {"disk",             std::make_shared<DataTypeString>()},
+        {"path",             std::make_shared<DataTypeString>()},
+        {"reason",           std::make_shared<DataTypeNullable>(std::make_shared<DataTypeString>())},
         {"min_block_number", std::make_shared<DataTypeNullable>(std::make_shared<DataTypeInt64>())},
         {"max_block_number", std::make_shared<DataTypeNullable>(std::make_shared<DataTypeInt64>())},
-        {"level", std::make_shared<DataTypeNullable>(std::make_shared<DataTypeUInt32>())}
+        {"level",            std::make_shared<DataTypeNullable>(std::make_shared<DataTypeUInt32>())}
     }});
     setInMemoryMetadata(storage_metadata);
 }
+static void calculateTotalSizeOnDiskImpl(const DiskPtr & disk, const String & from, UInt64 & total_size)
+{
+    /// Files or directories of detached part may not exist. Only count the size of existing files.
```

Comment:
According to the comment it should be enough to check `disk->exists` instead of try catch. Adding try catch protects against concurrent removal of files (btw `isFile` check below is not protected but can also throw exception as file might be removed too). May be only check with `exists`? (As the result will anyway be inconsistent in case of concurrent removal, so exception in this case is probably ok). 

Also it is better to substitute `listFiles` with `iterateDirectory` below.

Reply:
> According to the comment it should be enough to check `disk->exists` instead of try catch. Adding try catch protects against concurrent removal of files (btw `isFile` check below is not protected but can also throw exception as file might be removed too). May be only check with `exists`? (As the result will anyway be inconsistent in case of concurrent removal, so exception in this case is probably ok).
> 
> Also it is better to substitute `listFiles` with `iterateDirectory` below.

I think that no matter whether you use `listFiles` or `iterateDirectory`, when you execute `getFileSize`, you cannot guarantee whether the file exists at this time, so try catch is used. If the file does not exist, `isFile` return false. This may be OK. Any ideas?